### ISSUE TITLE
Amplify Pipeline-Generated Group Authorization Resolvers Broken

### ISSUE THREAD

**Issue by samjett247** [aws-amplify/amplify-cli#9977]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

14.17.1

### Amplify CLI Version

7.6.23

### What operating system are you using?

Mac

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

No Manual changes made

### Amplify Categories

api

### Amplify Commands

push

### Describe the bug

The bug is in the generation of the Mutation.update<TYPE>.auth.1.res.vtl pipeline resolver file when the "groups" authorization strategy is used. We define the following type inside our application:

```
type task
  @model
  @auth(rules: [{
        allow: groups
        groups: ["MyPermission"]
        operations: [update]
      }])
{
    fieldname: ID
      @primaryKey
    fieldrequired: String!
    fieldoptional: String
}
```

This will generate a file named `Mutation.updateTask.auth.1.res.vtl` in the api/build/resolvers directory. However, this file will vary based on the version of the amplify CLI (7.6.20 vs 7.6.23). The older version has the correct VTL file, the new version will generate an incorrect VTL file that will consider each user as Unauthorized (regardless of their permissions). 

### With Version 7.6.20
```vtl
## [Start] Authorization Steps. **
$util.qr($ctx.stash.put("hasAuth", true))
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#set( $nullAllowedFields = [] )
#set( $deniedFields = {} )
#if( $util.authType() == "User Pool Authorization" )
  #if( !$isAuthorized )
    #set( $staticGroupRoles = [{"claim":"cognito:groups","entity":"MyPermission","allowedFields":[],"nullAllowedFields":[]}] )
    #foreach( $groupRole in $staticGroupRoles )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get($groupRole.claim), []) )
      #if( $groupsInToken.contains($groupRole.entity) )
        #if( !$groupRole.allowedFields.isEmpty() || !$groupRole.nullAllowedFields.isEmpty() )
          $util.qr($allowedFields.addAll($groupRole.allowedFields))
          $util.qr($nullAllowedFields.addAll($groupRole.nullAllowedFields))
        #else
          #set( $isAuthorized = true )
          #break
        #end
      #end
    #end
  #end
#end
#if( !$isAuthorized && $allowedFields.isEmpty() && $nullAllowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #foreach( $entry in $util.map.copyAndRetainAllKeys($ctx.args.input, $inputFields).entrySet() )
    #if( $util.isNull($entry.value) && !$nullAllowedFields.contains($entry.key) )
      $util.qr($deniedFields.put($entry.key, ""))
    #end
  #end
  #foreach( $deniedField in $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
    $util.qr($deniedFields.put($deniedField, ""))
  #end
#end
#if( $deniedFields.keySet().size() > 0 )
  $util.error("Unauthorized on ${deniedFields.keySet()}", "Unauthorized")
#end
$util.toJson({})
## [End] Authorization Steps. **
```

### With Amplify CLI version 7.6.23

```vtl
## [Start] Authorization Steps. **
$util.qr($ctx.stash.put("hasAuth", true))
#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#set( $nullAllowedFields = [] )
#set( $deniedFields = {} )
#if( $util.authType() == "User Pool Authorization" )
  #if( !$isAuthorized )
    #set( $staticGroupRoles = [{"claim":"cognito:groups","entity":"MyPermission","allowedFields":[],"nullAllowedFields":[],"areAllFieldsAllowed":true,"areAllFieldsNullAllowed":false}] )
    #foreach( $groupRole in $staticGroupRoles )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get($groupRole.claim), []) )
      #if( $groupsInToken.contains($groupRole.entity) )
        #if( $groupRole.areAllFieldsAllowed && $groupRole.areAllFieldsNullAllowed )
          #set( $isAuthorized = true )
          #break
        #else
          $util.qr($allowedFields.addAll($groupRole.allowedFields))
          $util.qr($nullAllowedFields.addAll($groupRole.nullAllowedFields))
        #end
      #end
    #end
  #end
#end
#if( !$isAuthorized && $allowedFields.isEmpty() && $nullAllowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #foreach( $entry in $util.map.copyAndRetainAllKeys($ctx.args.input, $inputFields).entrySet() )
    #if( $util.isNull($entry.value) && !$nullAllowedFields.contains($entry.key) )
      $util.qr($deniedFields.put($entry.key, ""))
    #end
  #end
  #foreach( $deniedField in $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
    $util.qr($deniedFields.put($deniedField, ""))
  #end
#end
#if( $deniedFields.keySet().size() > 0 )
  $util.error("Unauthorized on ${deniedFields.keySet()}", "Unauthorized")
#end
$util.toJson({})
## [End] Authorization Steps. **
```

The reason the 7.6.23 template is invalid is that the "areAllFieldsNullAllowed" is always false, so to my understanding the case that sets the user to authorized in the 7.6.20 vtl file does not get hit. 

### Expected behavior

I expect to be able to update the `task` type if I have a "MyPermission" tag in my groups, regardless of what contents the record has. 

### Reproduction steps

See above. Run `amplify api gql-compile` with CLI versions 7.6.20 and then again with 7.6.23. 

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line


```

</details>

### Log output

<details>

```
type task
  @model
  @auth(rules: [{
        allow: groups
        groups: ["MyPermission"]
        operations: [update]
      }])
{
    fieldname: ID
      @primaryKey
    fieldrequired: String!
    fieldoptional: String
}
```

</details>

**Comment by samjett247** [comment]
Possibly - However I note that #9788 seems to be happening with 7.6.20, and here I am noting that 7.6.20 works for my use case but a higher upgrade (7.6.23) _creates this bug_. 

 #9370 seems to be more broad than this ticket but yes probably contains this scope.

**Comment by dmost714** [comment]
**TL;DR**
I just encountered this issue.
My code worked, then stopped working after amplify-cli update and a push.
Reverting to 7.6.20 (w/ amplify push -y) restores the code to working again.
Many many thanks to @josefaidt for pointing me to #9370.

**Details**
With the latest CLI, bug behavior is:
I have a "Campaigns" table.
I can create a new Campaign.
I can view the just created campaign.
But I can NOT update/mutate the campaign.
The record is crated correctly in DynamoDB. The 'group' field is set, and the current signed-in user belongs to that group.

Authorization modes:
- Default: Amazon Cognito User Pool
- IAM

`amplify status api -acm Campaign`
```
userPools:dynamicGroup:group
  ┌────────────────────┬────────┬──────┬────────┬────────┐
  │      (index)       │ create │ read │ update │ delete │
  ├────────────────────┼────────┼──────┼────────┼────────┤
  │         id         │  true  │ true │  true  │ false  │
  │       group        │  true  │ true │  true  │ false  │
  └────────────────────┴────────┴──────┴────────┴────────┘

```
The failure presents as an auth error:
```
"errors": [
    {
      "path": [
        "updateCampaign"
      ],
      "data": null,
      "errorType": "Unauthorized",
      "errorInfo": null,
      "locations": [
        {
          "line": 13,
          "column": 3,
          "sourceName": null
        }
      ],
      "message": "Not Authorized to access updateCampaign on type Mutation"
    }
  ]

```

The issue is reproducible from my React App as well as GraphiQL when signed in via user pools.


Here is a snippet from my schema:
```
type Campaign @model @auth(rules:[
  { allow: groups, groupsField: "group", operations: [create, read, update] },
  { allow: private, provider: iam }
])
{
  id: ID!
  group: String!
...
}

```

**Comment by lazpavel** [comment]
Hi @samjett247 I believe the issue you're having should be fixed by this PR once it gets released: https://github.com/aws-amplify/amplify-cli/pull/9859
